cli/src/merge_tools/builtin.rs

```diff
@@ -0,0 +1,588 @@
+                    num_bytes: old_num_bytes,
+                },
+                FileContents::Text {
+                    contents: _,
+                    hash: new_hash,
+                    num_bytes: new_num_bytes,
+                }
+                | FileContents::Binary {
+                    hash: new_hash,
+                    num_bytes: new_num_bytes,
+                },
+            ) => sections.push(scm_record::Section::Binary {
+                is_checked: false,
+                old_description: Some(Cow::Owned(describe_binary(
+                    old_hash.as_deref(),
+                    old_num_bytes,
+                ))),
+                new_description: Some(Cow::Owned(describe_binary(
+                    new_hash.as_deref(),
+                    new_num_bytes,
+                ))),
+            }),
+
+            (FileContents::Binary { hash, num_bytes }, FileContents::Absent) => {
+                sections.push(scm_record::Section::Binary {
+                    is_checked: false,
+                    old_description: Some(Cow::Owned(describe_binary(hash.as_deref(), num_bytes))),
+                    new_description: None,
+                })
+            }
+        }
+
+        files.push(scm_record::File {
+            old_path: None,
+            path: Cow::Owned(changed_path.to_fs_path(Path::new(""))),
+            file_mode: None,
+            sections,
+        });
+    }
+    Ok(files)
+}
+
+pub fn apply_diff_builtin(
+    store: Arc<Store>,
+    left_tree: &MergedTree,
+    right_tree: &MergedTree,
+    changed_files: Vec<RepoPath>,
+    files: &[scm_record::File],
+) -> Result<MergedTreeId, BackendError> {
+    let mut tree_builder =
+        MergedTreeBuilder::new(left_tree.store().clone(), left_tree.id().clone());
+    assert_eq!(
+        changed_files.len(),
+        files.len(),
+        "result had a different number of files"
+    );
+    for (path, file) in changed_files.into_iter().zip(files) {
+        let (selected, _unselected) = file.get_selected_contents();
+        match selected {
+            scm_record::SelectedContents::Absent => {
+                tree_builder.set_or_remove(path, Merge::absent());
+            }
+            scm_record::SelectedContents::Unchanged => {
+                // Do nothing.
+            }
+            scm_record::SelectedContents::Binary {
+                old_description: _,
+                new_description: _,
+            } => {
+                let value = right_tree.path_value(&path);
+                match value.into_resolved() {
+                    Ok(None) => {
+                        tree_builder.set_or_remove(path, Merge::absent());
+                    }
+                    Ok(Some(value)) => {
+                        tree_builder.set_or_remove(path, Merge::normal(value));
+                    }
+                    Err(merge) => {
+                        tree_builder.set_or_remove(path, merge);
+                    }
+                }
```

Comment:
This seems to always take the value from the right tree, so just `tree_builder.set_or_remove(path, value);` should work.